File: packages/sdk/src/modules/token.ts
import { ConfidentialTransfers } from "../artifacts/typechain/src/ConfidentialTransfers.js"
import { ConfidentialOFT } from "../artifacts/typechain/src/ConfidentialOFT.js"
import { ConfidentialTransfersBridgeable } from "../artifacts/typechain/src/ConfidentialTransfersBridgeable.js"
import { Proofs } from "./proofs.js"
import { SDKOptions, Target } from "./types.js"

export class Token extends Proofs {
  constructor(
    readonly token:
      | ConfidentialTransfers
      | ConfidentialTransfersBridgeable
      | ConfidentialOFT,
    options: SDKOptions,
  ) {
    super(options)
  }

  /* MISC */

  async сBalanceOf(address: string, cPrivateKey: bigint): Promise<bigint> {
    const accountData = await this.token.getAccount(address)
    return await this.decryptAmount(
      cPrivateKey,
      accountData.state.nonce,
      BigInt(accountData.state.eAmount),
    )
  }

  async generateOTK(key: bigint, nonce: bigint): Promise<bigint> {
    const { chainId, contractAddress } = await this.getTarget()
    return await Proofs.generateOTK(key, nonce, chainId, contractAddress)
  }

  async decryptAmount(
    key: bigint,
    nonce: bigint,
    eAmount: bigint,
  ): Promise<bigint> {
    const { chainId, contractAddress } = await this.getTarget()
    return await Proofs.decryptAmount(
      key,
      nonce,
      eAmount,
      chainId,
      contractAddress,
    )
  }

  async getTarget(): Promise<Target> {
    const network = await this.token.runner?.provider?.getNetwork()
    if (!network) throw new Error("Network not found")
    return {
      chainId: BigInt(network.chainId),
      contractAddress: BigInt(await this.token.getAddress()),
    }
  }
}

File: packages/go-sdk/sdk/token.go
package sdk

import (
	"context"
	"math/big"

	"github.com/ethereum/go-ethereum/common"
)

func (s *SDK) CBalanceOf(address common.Address, cPrivateKey *big.Int) (*big.Int, error) {
	accountData, err := s.Token.GetAccount(nil, address)
	if err != nil {
		return nil, err
	}

	return s.DecryptAmount(
		cPrivateKey,
		accountData.State.Nonce,
		accountData.State.EAmount,
	)
}

func (s *SDK) GenerateOTK(key, nonce *big.Int) (*big.Int, error) {
	target := s.GetTarget()
	return GenerateOTK(key, nonce, target.ChainId, target.ContractAddress)
}

func (s *SDK) DecryptAmount(key, nonce, eAmount *big.Int) (*big.Int, error) {
	target := s.GetTarget()
	return DecryptAmount(key, nonce, eAmount, target.ChainId, target.ContractAddress)
}

func (s *SDK) GetTarget() Target {
	chainId, err := s.Client.ChainID(context.Background())
	if err != nil {
		panic(err)
	}
	return Target{
		ChainId:         chainId,
		ContractAddress: s.TokenAddress.Big(),
	}
}
